software/controller/lib/hal/uart_base.cpp

```diff
@@ -310,15 +314,24 @@ void Channel::UARTInterruptHandler() {
   // Check for character match interrupt and trigger character match callback
   if (uart->status.bitfield.char_match) {
     uart->interrupt_clear.bitfield.char_match_clear = 1;
-    if (rx_listener_) rx_listener_->on_character_match();
+    OnCharacterMatch();
   }
 
   // Check for tx_complete interrupt to trigger callback
   if (uart->status.bitfield.tx_complete) {
     uart->interrupt_clear.bitfield.tx_complete_clear = 1;
     uart->control_reg1.bitfield.tx_complete_interrupt = 0;
-    if (tx_listener_) tx_listener_->on_tx_complete();
+    OnTxComplete();
   }
 }
 
+void Channel::OnTxComplete() {
+  tx_in_progress_ = false;
+  if (tx_listener_) tx_listener_->on_tx_complete();
```

Comment:
Here and several other places I suggest to add {}
if (tx_listener_) {
  tx_listener_->on_tx_complete();
}